In [1]:
from sqlalchemy import create_engine
import psycopg2
import pandas as pd

In [2]:
# Correct connection string for SQLAlchemy
conn_string = "postgresql://bigdata_jchai:bigdata_password8075jcci@postgres-warehouse:5432/financial_stock_dw"

# Create SQLAlchemy engine
db = create_engine(conn_string)
psql_conn = db.connect()

# Psycopg2 connection (still fine)
conn = psycopg2.connect(
    dbname="financial_stock_dw",
    user="bigdata_jchai",
    password="bigdata_password8075jcci",
    host="postgres-warehouse",
    port="5432"
)

conn.autocommit = True
cursor = conn.cursor()

In [3]:
engine = create_engine(conn_string)

In [4]:
pd.read_sql('select * from raw_finance_stock.finance_stock_sp500_price_hist', engine)

Date        Open        High         Low  \
0     2025-10-27 04:00:00+00:00  189.990005  192.000000  188.429993   
1     2025-10-28 04:00:00+00:00  193.050003  203.149994  191.910004   
2     2025-10-29 04:00:00+00:00  207.979996  212.190002  204.779999   
3     2025-10-30 04:00:00+00:00  205.149994  206.160004  201.410004   
4     2025-10-31 04:00:00+00:00  206.449997  207.970001  202.070007   
...                         ...         ...         ...         ...   
11060 2025-11-19 05:00:00+00:00   28.719999   28.740000   28.350000   
11061 2025-11-20 05:00:00+00:00   28.610001   28.780001   27.920000   
11062 2025-11-21 05:00:00+00:00   28.100000   29.100000   28.100000   
11063 2025-11-24 05:00:00+00:00   29.000000   29.160000   28.620001   
11064 2025-11-25 05:00:00+00:00   28.504999   29.240000   28.504999   

            Close     Volume  Dividends  Stock Splits Symbol  \
0      191.490005  153452700        0.0           0.0   NVDA   
1      201.029999  297986200        0.0           0.0   NVDA   
2      207.039993  308829600        0.0           0.0   NVDA   
3      202.889999  178864400        0.0           0.0   NVDA   
4      202.490005  179802200        0.0           0.0   NVDA   
...           ...        ...        ...           ...    ...   
11060   28.360001     670900        0.0           0.0    NWS   
11061   27.980000     755900        0.0           0.0    NWS   
11062   29.000000     767900        0.0           0.0    NWS   
11063   28.860001     908700        0.0           0.0    NWS   
11064   28.700001      61876        0.0           0.0    NWS   

                         Ingested_Time  
0     2025-11-25 14:56:04.196316+00:00  
1     2025-11-25 14:56:04.196316+00:00  
2     2025-11-25 14:56:04.196316+00:00  
3     2025-11-25 14:56:04.196316+00:00  
4     2025-11-25 14:56:04.196316+00:00  
...                                ...  
11060 2025-11-25 14:56:04.196316+00:00  
11061 2025-11-25 14:56:04.196316+00:00  
11062 2025-11-25 14:56:04.196316+00:00  
11063 2025-11-25 14:56:04.196316+00:00  
11064 2025-11-25 14:56:04.196316+00:00  

[11065 rows x 10 columns]

🏷 Field-Level Metadata
1. Date

Type: timestamp with timezone

Description: Trading date and time for each price record.

Example: 2025-10-27 04:00:00+00:00

2. Open

Type: float

Description: Opening stock price at the beginning of the trading day.

Example: 189.99

3. High

Type: float

Description: Highest traded price during the day.

Example: 192.00

4. Low

Type: float

Description: Lowest traded price during the day.

Example: 188.42

5. Close

Type: float

Description: Closing stock price at the end of the trading session.

Example: 191.49

6. Volume

Type: bigint

Description: Total number of shares traded during the day.

Example: 153452700

7. Dividends

Type: float

Description: Dividend value paid on that date (per share).
Usually 0.0 on non-dividend days.

Example: 0.0

8. Stock Splits

Type: float

Description: Ratio representing split events.
0.0 means no split; e.g., 2.0 means a 2-for-1 split.

Example: 0.0

9. Symbol

Type: varchar

Description: The company ticker symbol representing the stock.

Example: NVDA

10. Ingested_Time

Type: timestamp with timezone

Description: Timestamp indicating when the data was ingested by the ETL pipeline.

Example: 2025-11-25 14:56:04.196316+00:00

In [5]:
df = pd.read_sql('select * from raw_finance_stock.finance_stock_sp500_profile', engine)
df

Exchange Symbol                        Shortname  \
0        NMS   NVDA               NVIDIA Corporation   
1        NMS   AAPL                       Apple Inc.   
2        NMS   MSFT            Microsoft Corporation   
3        NMS   AMZN                 Amazon.com, Inc.   
4        NMS  GOOGL                    Alphabet Inc.   
..       ...    ...                              ...   
498      NYQ    MOH            Molina Healthcare Inc   
499      NMS    LKQ                  LKQ Corporation   
500      NMS   SOLS  Solstice Advanced Materials Inc   
501      NYQ    MHK          Mohawk Industries, Inc.   
502      NMS    NWS                 News Corporation   

                              Longname                  Sector  \
0                   NVIDIA Corporation              Technology   
1                           Apple Inc.              Technology   
2                Microsoft Corporation              Technology   
3                     Amazon.com, Inc.       Consumer Cyclical   
4                        Alphabet Inc.  Communication Services   
..                                 ...                     ...   
498            Molina Healthcare, Inc.              Healthcare   
499                    LKQ Corporation       Consumer Cyclical   
500  Solstice Advanced Materials, Inc.         Basic Materials   
501            Mohawk Industries, Inc.       Consumer Cyclical   
502                   News Corporation  Communication Services   

                               Industry  Current_Price     Market_Cap  \
0                        Semiconductors         177.00  4301985153024   
1                  Consumer Electronics         278.85  4138242670592   
2             Software - Infrastructure         492.01  3657192177664   
3                       Internet Retail         233.22  2493172154368   
4        Internet Content & Information         320.18  3878058983424   
..                                  ...            ...            ...   
498                    Healthcare Plans         148.26     8035691520   
499                          Auto Parts          29.69     7639034880   
500                 Specialty Chemicals          47.68     7568124928   
501  Furnishings, Fixtures & Appliances         115.90     7201088512   
502                       Entertainment          29.43    16430740480   

     Net_Income_To_Common        Ebitda  Revenue_Growth  \
0            9.919800e+10  1.126960e+11           0.625   
1            1.120100e+11  1.447480e+11           0.079   
2            1.049120e+11  1.664370e+11           0.184   
3            7.648200e+10  1.396970e+11           0.134   
4            1.242510e+11  1.451740e+11           0.159   
..                    ...           ...             ...   
498          8.830000e+08  1.457000e+09           0.116   
499          6.960000e+08  1.640000e+09           0.013   
500          3.290000e+08  9.850000e+08           0.068   
501          4.211000e+08  1.321300e+09           0.014   
502          4.740000e+08  1.148000e+09           0.023   

                       Ingested_Time  
0   2025-11-29 18:30:17.348340+00:00  
1   2025-11-29 18:30:17.348340+00:00  
2   2025-11-29 18:30:17.348340+00:00  
3   2025-11-29 18:30:17.348340+00:00  
4   2025-11-29 18:30:17.348340+00:00  
..                               ...  
498 2025-11-29 18:30:17.348340+00:00  
499 2025-11-29 18:30:17.348340+00:00  
500 2025-11-29 18:30:17.348340+00:00  
501 2025-11-29 18:30:17.348340+00:00  
502 2025-11-29 18:30:17.348340+00:00  

[503 rows x 12 columns]

In [6]:
# Add auto-increment PK
df["Profile_Key"] = range(1, len(df) + 1)
# Reorder columns
column_order = [
    "Profile_Key", "Exchange", "Symbol", "Shortname", "Longname", 
    "Sector", "Industry", "Current_Price", "Market_Cap", 
    "Net_Income_To_Common", "Ebitda", "Revenue_Growth", "Ingested_Time"
]

df = df[column_order]
df

Profile_Key Exchange Symbol                        Shortname  \
0              1      NMS   NVDA               NVIDIA Corporation   
1              2      NMS   AAPL                       Apple Inc.   
2              3      NMS   MSFT            Microsoft Corporation   
3              4      NMS   AMZN                 Amazon.com, Inc.   
4              5      NMS  GOOGL                    Alphabet Inc.   
..           ...      ...    ...                              ...   
498          499      NYQ    MOH            Molina Healthcare Inc   
499          500      NMS    LKQ                  LKQ Corporation   
500          501      NMS   SOLS  Solstice Advanced Materials Inc   
501          502      NYQ    MHK          Mohawk Industries, Inc.   
502          503      NMS    NWS                 News Corporation   

                              Longname                  Sector  \
0                   NVIDIA Corporation              Technology   
1                           Apple Inc.              Technology   
2                Microsoft Corporation              Technology   
3                     Amazon.com, Inc.       Consumer Cyclical   
4                        Alphabet Inc.  Communication Services   
..                                 ...                     ...   
498            Molina Healthcare, Inc.              Healthcare   
499                    LKQ Corporation       Consumer Cyclical   
500  Solstice Advanced Materials, Inc.         Basic Materials   
501            Mohawk Industries, Inc.       Consumer Cyclical   
502                   News Corporation  Communication Services   

                               Industry  Current_Price     Market_Cap  \
0                        Semiconductors         177.00  4301985153024   
1                  Consumer Electronics         278.85  4138242670592   
2             Software - Infrastructure         492.01  3657192177664   
3                       Internet Retail         233.22  2493172154368   
4        Internet Content & Information         320.18  3878058983424   
..                                  ...            ...            ...   
498                    Healthcare Plans         148.26     8035691520   
499                          Auto Parts          29.69     7639034880   
500                 Specialty Chemicals          47.68     7568124928   
501  Furnishings, Fixtures & Appliances         115.90     7201088512   
502                       Entertainment          29.43    16430740480   

     Net_Income_To_Common        Ebitda  Revenue_Growth  \
0            9.919800e+10  1.126960e+11           0.625   
1            1.120100e+11  1.447480e+11           0.079   
2            1.049120e+11  1.664370e+11           0.184   
3            7.648200e+10  1.396970e+11           0.134   
4            1.242510e+11  1.451740e+11           0.159   
..                    ...           ...             ...   
498          8.830000e+08  1.457000e+09           0.116   
499          6.960000e+08  1.640000e+09           0.013   
500          3.290000e+08  9.850000e+08           0.068   
501          4.211000e+08  1.321300e+09           0.014   
502          4.740000e+08  1.148000e+09           0.023   

                       Ingested_Time  
0   2025-11-29 18:30:17.348340+00:00  
1   2025-11-29 18:30:17.348340+00:00  
2   2025-11-29 18:30:17.348340+00:00  
3   2025-11-29 18:30:17.348340+00:00  
4   2025-11-29 18:30:17.348340+00:00  
..                               ...  
498 2025-11-29 18:30:17.348340+00:00  
499 2025-11-29 18:30:17.348340+00:00  
500 2025-11-29 18:30:17.348340+00:00  
501 2025-11-29 18:30:17.348340+00:00  
502 2025-11-29 18:30:17.348340+00:00  

[503 rows x 13 columns]

In [24]:
import yfinance as yf

In [25]:
def get_sp500_list():
    # Get the list of S&P 500 companies from a reliable source
    url = "https://www.slickcharts.com/sp500"
    df = pd.read_html(url)[0]
    df['Symbol'] = df['Symbol'].str.replace('.', '-', regex=False)  # fix BRK.B → BRK-B
    sp500_list = df['Symbol'].values.tolist()
    print("S&P 500 company list extracted successfully.")
    print(f"Total companies in S&P 500: {len(sp500_list)}")
    return sp500_list

# try:
#     # Get S&P 500 company list
#     sp500_list = get_sp500_list()

#     # Initialize an empty list to store data
#     data_list = []

#     for symbol in sp500_list:
#         ticker = yf.Ticker(symbol)
#         info = ticker.info  # Full metadata

#         data_list.append({
#                 "Exchange": info.get("exchange"),
#                 "Symbol": symbol,
#                 "Shortname": info.get("shortName"),
#                 "Longname": info.get("longName"),
#                 "Sector": info.get("sector"),
#                 "Industry": info.get("industry"),
#                 "Current_Price": info.get("currentPrice"),
#                 "Market_Cap": info.get("marketCap"),
#                 "Net_Income_To_Common": info.get("netIncomeToCommon"),
#                 "Ebitda": info.get("ebitda"),
#                 "Revenue_Growth": info.get("revenueGrowth"),
#             })

#         sp_profile_df = pd.DataFrame(data_list)
# except Exception as e:
#     print(f"The ERROR is: {e}")


# S&P 500 index ticker on Yahoo Finance
try:
    # Get S&P 500 company list
    sp500_list = get_sp500_list()

    # Initialize an empty list to store data
    data_list = []

    for symbol in sp500_list:
        ticker = yf.Ticker(symbol)
        hist = ticker.history(period='1mo', interval='1d', start=None, end=None)  # Historical price data

        hist.reset_index(inplace=True)
        hist['Symbol'] = symbol  # Add symbol column

        data_list.append(hist)

    sp_price_history_df = pd.concat(data_list, ignore_index=True)
except Exception as e:
    print(f"The ERROR is: {e}")

S&P 500 company list extracted successfully.
Total companies in S&P 500: 503
S&P 500 company list extracted successfully.
Total companies in S&P 500: 503


Date        Open        High         Low  \
0     2025-10-29 00:00:00-04:00  207.979996  212.190002  204.779999   
1     2025-10-30 00:00:00-04:00  205.149994  206.160004  201.410004   
2     2025-10-31 00:00:00-04:00  206.449997  207.970001  202.070007   
3     2025-11-03 00:00:00-05:00  208.080002  211.339996  205.559998   
4     2025-11-04 00:00:00-05:00  203.000000  203.970001  197.929993   
...                         ...         ...         ...         ...   
11060 2025-11-21 00:00:00-05:00   28.100000   29.100000   28.100000   
11061 2025-11-24 00:00:00-05:00   29.000000   29.160000   28.620001   
11062 2025-11-25 00:00:00-05:00   29.020000   29.240000   28.510000   
11063 2025-11-26 00:00:00-05:00   29.139999   29.379999   28.990000   
11064 2025-11-28 00:00:00-05:00   29.330000   29.549999   29.240000   

            Close     Volume  Dividends  Stock Splits Symbol  
0      207.039993  308829600        0.0           0.0   NVDA  
1      202.889999  178864400        0.0           0.0   NVDA  
2      202.490005  179802200        0.0           0.0   NVDA  
3      206.880005  180267300        0.0           0.0   NVDA  
4      198.690002  188919300        0.0           0.0   NVDA  
...           ...        ...        ...           ...    ...  
11060   29.000000     767900        0.0           0.0    NWS  
11061   28.860001     910100        0.0           0.0    NWS  
11062   29.059999     642200        0.0           0.0    NWS  
11063   29.320000     640000        0.0           0.0    NWS  
11064   29.430000     776100        0.0           0.0    NWS  

[11065 rows x 9 columns]

In [26]:
sp_price_history_df

Date        Open        High         Low  \
0     2025-10-29 00:00:00-04:00  207.979996  212.190002  204.779999   
1     2025-10-30 00:00:00-04:00  205.149994  206.160004  201.410004   
2     2025-10-31 00:00:00-04:00  206.449997  207.970001  202.070007   
3     2025-11-03 00:00:00-05:00  208.080002  211.339996  205.559998   
4     2025-11-04 00:00:00-05:00  203.000000  203.970001  197.929993   
...                         ...         ...         ...         ...   
11060 2025-11-21 00:00:00-05:00   28.100000   29.100000   28.100000   
11061 2025-11-24 00:00:00-05:00   29.000000   29.160000   28.620001   
11062 2025-11-25 00:00:00-05:00   29.020000   29.240000   28.510000   
11063 2025-11-26 00:00:00-05:00   29.139999   29.379999   28.990000   
11064 2025-11-28 00:00:00-05:00   29.330000   29.549999   29.240000   

            Close     Volume  Dividends  Stock Splits Symbol  
0      207.039993  308829600        0.0           0.0   NVDA  
1      202.889999  178864400        0.0           0.0   NVDA  
2      202.490005  179802200        0.0           0.0   NVDA  
3      206.880005  180267300        0.0           0.0   NVDA  
4      198.690002  188919300        0.0           0.0   NVDA  
...           ...        ...        ...           ...    ...  
11060   29.000000     767900        0.0           0.0    NWS  
11061   28.860001     910100        0.0           0.0    NWS  
11062   29.059999     642200        0.0           0.0    NWS  
11063   29.320000     640000        0.0           0.0    NWS  
11064   29.430000     776100        0.0           0.0    NWS  

[11065 rows x 9 columns]

In [ ]:
        "Officers_0_name": info.get("companyOfficers")[0].get("name"),
        "Officers_0_title": info.get("companyOfficers")[0].get("title"),
        "Officers_0_yearBorn": info.get("companyOfficers")[0].get("yearBorn"),
        "Officers_0_fiscalYear": info.get("companyOfficers")[0].get("fiscalYear"),
        "Officers_0_totalPay": info.get("companyOfficers")[0].get("totalPay"),
        "Officers_1_name": info.get("companyOfficers")[1].get("name"),
        "Officers_1_title": info.get("companyOfficers")[1].get("title"),
        "Officers_1_yearBorn": info.get("companyOfficers")[1].get("yearBorn"),
        "Officers_1_fiscalYear": info.get("companyOfficers")[1].get("fiscalYear"),
        "Officers_1_totalPay": info.get("companyOfficers")[1].get("totalPay"),
        "Officers_2_name": info.get("companyOfficers")[2].get("name"),
        "Officers_2_title": info.get("companyOfficers")[2].get("title"),
        "Officers_2_yearBorn": info.get("companyOfficers")[2].get("yearBorn"),
        "Officers_2_fiscalYear": info.get("companyOfficers")[2].get("fiscalYear"),
        "Officers_2_totalPay": info.get("companyOfficers")[2].get("totalPay"),
        "Officers_3_name": info.get("companyOfficers")[3].get("name"),
        "Officers_3_title": info.get("companyOfficers")[3].get("title"),
        "Officers_3_yearBorn": info.get("companyOfficers")[3].get("yearBorn"),
        "Officers_3_fiscalYear": info.get("companyOfficers")[3].get("fiscalYear"),
        "Officers_3_totalPay": info.get("companyOfficers")[3].get("totalPay"),
        "Officers_4_name": info.get("companyOfficers")[4].get("name"),
        "Officers_4_title": info.get("companyOfficers")[4].get("title"),
        "Officers_4_yearBorn": info.get("companyOfficers")[4].get("yearBorn"),
        "Officers_4_fiscalYear": info.get("companyOfficers")[4].get("fiscalYear"),
        "Officers_4_totalPay": info.get("companyOfficers")[4].get("totalPay"),
        "Officers_5_name": info.get("companyOfficers")[5].get("name"),
        "Officers_5_title": info.get("companyOfficers")[5].get("title"),
        "Officers_5_yearBorn": info.get("companyOfficers")[5].get("yearBorn"),
        "Officers_5_fiscalYear": info.get("companyOfficers")[5].get("fiscalYear"),
        "Officers_5_totalPay": info.get("companyOfficers")[5].get("totalPay"),
        "Officers_6_name": info.get("companyOfficers")[6].get("name"),
        "Officers_6_title": info.get("companyOfficers")[6].get("title"),
        "Officers_6_yearBorn": info.get("companyOfficers")[6].get("yearBorn"),
        "Officers_6_fiscalYear": info.get("companyOfficers")[6].get("fiscalYear"),
        "Officers_6_totalPay": info.get("companyOfficers")[6].get("totalPay"),
        "Officers_7_name": info.get("companyOfficers")[7].get("name"),
        "Officers_7_title": info.get("companyOfficers")[7].get("title"),
        "Officers_7_yearBorn": info.get("companyOfficers")[7].get("yearBorn"),
        "Officers_7_fiscalYear": info.get("companyOfficers")[7].get("fiscalYear"),
        "Officers_7_totalPay": info.get("companyOfficers")[7].get("totalPay"),
        "Officers_8_name": info.get("companyOfficers")[8].get("name"),
        "Officers_8_title": info.get("companyOfficers")[8].get("title"),
        "Officers_8_yearBorn": info.get("companyOfficers")[8].get("yearBorn"),
        "Officers_8_fiscalYear": info.get("companyOfficers")[8].get("fiscalYear"),
        "Officers_8_totalPay": info.get("companyOfficers")[8].get("totalPay"),
        "Officers_9_name": info.get("companyOfficers")[9].get("name"),
        "Officers_9_title": info.get("companyOfficers")[9].get("title"),
        "Officers_9_yearBorn": info.get("companyOfficers")[9].get("yearBorn"),
        "Officers_9_fiscalYear": info.get("companyOfficers")[9].get("fiscalYear"),
        "Officers_9_totalPay": info.get("companyOfficers")[9].get("totalPay"),